### ISSUE TITLE
`erf` produces different fp32 result for the same input based on tensor shape

### ISSUE THREAD

**Issue by wuyii8941** [openxla/xla#41122]
## Summary

XLA's `erf` lowering is not shape-invariant: the same fp32 input produces a different output depending on the innermost axis size of the enclosing tensor. For `x = 4.5`:

- Shape with innermost dim **= 1** (e.g. `[1]`, `[7, 64, 1]`) → `erf(x) = 0.9999998212`
- Shape with innermost dim **≥ 2** (e.g. `[2]`, `[7, 64, 5]`) → `erf(x) = 1.0` (correctly rounded)

The two outputs differ by ~3 fp32 ULPs. A single downstream `floor` / `ceil` / `>=` call amplifies this into an integer-valued, silent divergence.

The bug reproduces identically in JAX, which confirms the root cause lives in XLA's `erf` HLO lowering (not in TensorFlow's XLA bridge).

## Versions tested

| stack | version | reproduces? |
|---|---|---|
| TensorFlow | 2.21.0 (stable, 2026-03-06) | ✓ |
| TensorFlow | tf-nightly `2.22.0-dev20260417` | ✓ |
| JAX | 0.9.2 | ✓ |

Linux x86_64; reproduced on both CPU-only and CUDA 12 / NVIDIA T4 configurations.

## Minimal reproducer (TensorFlow, self-contained)

```python
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
import tensorflow as tf

@tf.function(jit_compile=True)
def xla_erf(x):
    return tf.math.erf(x)

print(f"TF version: {tf.__version__}")
for shape in [(1,), (2,), (7, 1), (7, 64, 1), (7, 64, 5), (1, 1, 1)]:
    x = tf.fill(shape, tf.constant(4.5, dtype=tf.float32))
    e  = tf.math.erf(x).numpy()
    xl = xla_erf(x).numpy()
    print(f"shape {str(shape):>13}  eager={float(e.flat[0]):.10f}  XLA={float(xl.flat[0]):.10f}  match={float(e.flat[0]) == float(xl.flat[0])}")
```

(The same result holds for `tf.raw_ops.Erf`.)

Expected (what `math.erf(4.5)` rounds to in fp32): `1.0` for every shape.
Observed on TF 2.21.0 and tf-nightly `2.22.0-dev20260417`:

```
shape          (1,)  eager=1.0000000000  XLA=0.9999998212  match=False
shape          (2,)  eager=1.0000000000  XLA=1.0000000000  match=True
shape        (7, 1)  eager=1.0000000000  XLA=0.9999998212  match=False
shape    (7, 64, 1)  eager=1.0000000000  XLA=0.9999998212  match=False
shape    (7, 64, 5)  eager=1.0000000000  XLA=1.0000000000  match=True
shape     (1, 1, 1)  eager=1.0000000000  XLA=0.9999998212  match=False
```

Exactly the same pattern appears when we bypass TF entirely and go through JAX:

## Cross-framework reproducer (JAX, confirms the bug is in XLA)

```python
import jax, jax.numpy as jnp
from jax import lax
import numpy as np

@jax.jit
def jit_erf(x):
    return lax.erf(x)

print(f"jax: {jax.__version__}")
for shape in [(1,), (2,), (7, 1), (7, 64, 1), (7, 64, 5), (1, 1, 1)]:
    x = jnp.full(shape, 4.5, dtype=jnp.float32)
    e  = np.asarray(lax.erf(x)).flat[0]
    xl = np.asarray(jit_erf(x)).flat[0]
    print(f"shape {str(shape):>13}  eager={float(e):.10f}  jit={float(xl):.10f}")
```

Output on jax 0.9.2:

```
shape          (1,)  eager=0.9999998212  jit=0.9999998212
shape          (2,)  eager=1.0000000000  jit=1.0000000000
shape        (7, 1)  eager=0.9999998212  jit=0.9999998212
shape    (7, 64, 1)  eager=0.9999998212  jit=0.9999998212
shape    (7, 64, 5)  eager=1.0000000000  jit=1.0000000000
shape     (1, 1, 1)  eager=0.9999998212  jit=0.9999998212
```

Every XLA-backed path (TF+jit, JAX-eager, JAX+jit) agrees on the shape-dependent output. Only TensorFlow's non-XLA Eigen kernel (used in TF eager) returns the correctly-rounded `1.0` for every shape.

| shape | TF eager (Eigen) | TF `jit_compile=True` (XLA) | JAX `lax.erf` (XLA) | JAX `jit(lax.erf)` (XLA) |
|---|---|---|---|---|
| `(1,)`        | `1.0000000000` | `0.9999998212` | `0.9999998212` | `0.9999998212` |
| `(2,)`        | `1.0000000000` | `1.0000000000` | `1.0000000000` | `1.0000000000` |
| `(7, 64, 1)`  | `1.0000000000` | `0.9999998212` | `0.9999998212` | `0.9999998212` |
| `(7, 64, 5)`  | `1.0000000000` | `1.0000000000` | `1.0000000000` | `1.0000000000` |

This rules out TF's XLA bridge as the source: same XLA, different frontend, same bug.

## Suggested fix

Apply the same `erf` polynomial that XLA already uses for innermost-dim ≥ 2 when the innermost dim is 1. There is no algorithmic reason the accuracy should drop for size-1 innermost tensors — the same approximation that produces `1.0` for shape `(2,)` can be applied to shape `(1,)`.

**Comment by kredd2506** [comment]
I traced the erf lowering to `xla/codegen/intrinsic/erf.cc` — `EmitErfF32` has a clamp at `|x| >= 3.83` that should return `1.0` for `x=4.5`. The CPU emitter in `xla/service/cpu/elemental_math_emitter.cc` also routes F32 erf through this same intrinsic.

So both scalar and vector paths seem to use the same clamp logic. Is the innermost dim=1 case bypassing this intrinsic entirely (e.g., falling back to a libm `erff` call), or could LLVM be optimizing the clamp differently in scalar mode?

**Comment by wuyii8941** [comment]
> I traced the erf lowering to `xla/codegen/intrinsic/erf.cc` — `EmitErfF32` has a clamp at `|x| >= 3.83` that should return `1.0` for `x=4.5`. The CPU emitter in `xla/service/cpu/elemental_math_emitter.cc` also routes F32 erf through this same intrinsic.我追踪到 erf 值降低的原因在 `xla/codegen/intrinsic/erf.cc` 中 `EmitErfF32` 在 `|x| >= 3.83` 处设置了钳位，对于 `x=4.5` 应该返回 `1.0` `xla/service/cpu/elemental_math_emitter.cc` 文件中的 CPU 发射器也通过这个内部函数处理 F32 erf 值。
> 
> So both scalar and vector paths seem to use the same clamp logic. Is the innermost dim=1 case bypassing this intrinsic entirely (e.g., falling back to a libm `erff` call), or could LLVM be optimizing the clamp differently in scalar mode?标量路径和向量路径似乎都使用了相同的钳位逻辑。最内层的 dim=1 情况是否完全绕过了这个内部函数（例如，回退到 libm `erff` 调用），或者 LLVM 是否在标量模式下对钳位进行了不同的优化？

  Hi, thanks for looking into this. The LLVM IR dumped with --xla_dump_to=/tmp/xla --xla_dump_hlo_as_text=true shows it's option (a) — the scalar path     
  bypasses EmitErfF32 entirely — but it's not falling back to libm erff. It's a distinct inline rational-approximation routine with a different clamp 
  constant, so LLVM never even sees a unified clamp to optimize.                                                                                           
                                                                                                                                                           
  Side-by-side on the .ir-with-opt.ll for the same Erf HLO, f32[4,1] vs f32[4,2]:                                                                          
                                                                                                                                                           
  f32[4,2] (vector path, matches EmitErfF32):                                                                                                              
  %11 = call <2 x float> @llvm.fabs.v2f32(<2 x float> %10)                         
  %12 = fcmp oge <2 x float> %11, splat (float 0x400EA8F960000000)  ; |x| >= 3.83251                                                                       
  %13 = fmul <2 x float> %10, %10                                     ; x²                                                                                 
  ; 4 FMAs: numerator poly on x²                                                                                                                           
  %18 = fmul <2 x float> %10, %17                                     ; x · P(x²)                                                                          
  ; 6 FMAs: denominator poly on x²                                                                                                                         
  %25 = call <2 x float> @llvm.copysign.v2f32(splat 1.0, %10)         ; sign(x)·1.0                                                                        
  %26 = fdiv <2 x float> %18, %24                                                                                                                          
  %27 = select <2 x i1> %12, <2 x float> %25, <2 x float> %26          ; saturation select                                                                 
  store <2 x float> %27                                                                                                                                    
                                                                                                                                                           
  f32[4,1] (scalar path, no saturation select):                                                                                                            
  %12 = tail call float @llvm.maximum.f32(%11, float 0xC00DF38D00000000)  ; input clamp to -3.74392
  %13 = tail call float @llvm.minimum.f32(%12, float 0x400DF38D00000000)  ;                 +3.74392                                                       
  %14 = fmul float %13, %13                                                                         
  ; 4 FMAs (numerator, f32 scalar)                                                                                                                         
  %19 = fmul float %13, %18                                                        
  ; 6 FMAs (denominator)                                                                                                                                   
  %26 = fdiv float %19, %25                                            ; polynomial result                                                                 
  store float %26                                                       ; no |x|>=threshold check, no select                                               
                                                                                                                                                           
  Two pieces of evidence this is emission-time divergence, not LLVM optimization:  
                                                                                                                                                           
  1. The saturation thresholds are already different in the IR. Vector path uses 0x400EA8F960000000 = 3.83251…; scalar path clamps to ±0x400DF38D00000000 =
   ±3.74392…. If both paths came from the same EmitErfF32, they would share the 3.83 constant — LLVM doesn't rewrite constants.                            
  2. The vector path emits fabs + fcmp + copysign + select as the saturation tail. The scalar path emits nothing after the fdiv. LLVM has no pass that     
  deletes a saturation select; it would have to be absent at emission.                                                                                     
                                                                                                                                                           
  So the flow must be that when XLA lowers an Erf fusion whose innermost dim is 1 (and therefore the LMHLO / loop emitter picks scalar rather than         
  vectorized LLVM types), the elemental dispatch takes a different branch than the one going through EmitErfF32 in elemental_math_emitter.cc. It looks like
   EmitErfF32 is reached only from the vectorized code path. For x = 4.5 the scalar path then evaluates the rational approximation at the clamped input
  3.74392, which in f32 gives 0.9999998212… — matching what we observe.                                                                                    
                                                                                   
  Minimal bit-level observation if useful:                                                                                                                 
  - 1.0 − 0.9999998212 = 1.788e-7 ≈ 3 · 2⁻²⁴ → exactly 3 f32 ULPs below 1.0        
  - Correctly-rounded f32 of math.erf(4.5) = 0.99999999980… is 1.0

**Comment by kredd2506** [comment]
Thanks — the IR dump pointed right to it. Grepping for the 3.7439 constant led me to [`xla/codegen/emitters/transforms/expand_float_ops.cc:85`](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L85), which looks like the second emitter you mentioned. Checking the [`kAlpha`/`kBeta` coefficients](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L71-L81) — as far as I can tell they appear to match the ones in [`erf.cc`](https://github.com/openxla/xla/blob/main/xla/codegen/intrinsic/erf.cc#L49-L70), so if I'm reading it right the polynomial itself might already be the same, and the difference is really in the saturation handling. I also noticed the [comment on line 83-84](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L83-L84) mentions *"outside of which x should be +/-1"*, but I couldn't find where that ±1 fallback actually happens in the [implementation](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L101-L107) — though I might be missing something.

If that reading is right, would a saturation select after [line 107](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L107) (comparing the original unclamped x against the threshold) be along the right lines, or am I oversimplifying?

On impact — I'm trying to build a mental model: GELU uses erf everywhere but the 3-ULP error seems like it might get absorbed by `0.5 * x * (1 + ...)`. The cases that worry me are direct CDF / probabilistic uses and anywhere `floor(erf)`, `ceil(erf)`, or comparisons against ±1 come up. Innermost-dim=1 seems uncommon in dense tensors but could be natural after `keepdims=True` reductions or per-channel parameters. Does any of that line up with what you've seen in practice, or am I off base?

**Comment by wuyii8941** [comment]
> Thanks — the IR dump pointed right to it. Grepping for the 3.7439 constant led me to [`xla/codegen/emitters/transforms/expand_float_ops.cc:85`](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L85), which looks like the second emitter you mentioned. Checking the [`kAlpha`/`kBeta` coefficients](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L71-L81) — as far as I can tell they appear to match the ones in [`erf.cc`](https://github.com/openxla/xla/blob/main/xla/codegen/intrinsic/erf.cc#L49-L70), so if I'm reading it right the polynomial itself might already be the same, and the difference is really in the saturation handling. I also noticed the [comment on line 83-84](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L83-L84) mentions _"outside of which x should be +/-1"_, but I couldn't find where that ±1 fallback actually happens in the [implementation](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L101-L107) — though I might be missing something.
> 
> If that reading is right, would a saturation select after [line 107](https://github.com/openxla/xla/blob/main/xla/codegen/emitters/transforms/expand_float_ops.cc#L107) (comparing the original unclamped x against the threshold) be along the right lines, or am I oversimplifying?
> 
> On impact — I'm trying to build a mental model: GELU uses erf everywhere but the 3-ULP error seems like it might get absorbed by `0.5 * x * (1 + ...)`. The cases that worry me are direct CDF / probabilistic uses and anywhere `floor(erf)`, `ceil(erf)`, or comparisons against ±1 come up. Innermost-dim=1 seems uncommon in dense tensors but could be natural after `keepdims=True` reductions or per-channel parameters. Does any of that line up with what you've seen in practice, or am I off base?

  Yeah, that's exactly right — the polynomial matches, it's just missing the ±1 fallback that the comment on 83-84 describes. A select after line 107  should do it.                                                                                                                                        
                                                                                                                                                      
  For impact: GELU is fine, like you said. We hit this through floor(erf(x)) where the input came out of a keepdims=True reduction (so innermost dim =  1). The whole tensor flipped from 1 to 0. That keepdims pattern is pretty natural in normalization and attention code, so it's not hard to run into.

**Comment by kredd2506** [comment]
Submitted a PR: #41167

Adds a saturation select in `expand_float_ops.cc` that checks the original unclamped x against the threshold, returning `copysign(1.0, x)` when |x| ≥ kErfInvOneMinusHalfULP. Matches the strategy used by `EmitErfF32` in `erf.cc`. All 19 tests in `//xla/codegen/emitters/transforms/tests:tests` pass locally.
